hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/DeleteOperation.java

```diff
@@ -0,0 +1,452 @@
+
+/**
+ * Implementation of the delete operation.
+ * For an authoritative S3Guarded store, after the list and delete of the
+ * combined store, we repeat against raw S3.
+ * This will correct for any situation where the authoritative listing is
+ * incomplete.
+ */
+public class DeleteOperation extends AbstractStoreOperation {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      DeleteOperation.class);
+
+  /**
+   * This is a switch to turn on when trying to debug
+   * deletion problems; it requests the results of
+   * the delete call from AWS then audits them.
+   */
+  private static final boolean AUDIT_DELETED_KEYS = true;
+
+  /**
+   * Used to stop any re-entrancy of the rename.
+   * This is an execute-once operation.
+   */
+  private final AtomicBoolean executed = new AtomicBoolean(false);
+
+  private final S3AFileStatus status;
+
+  private final boolean recursive;
+
+  private final OperationCallbacks callbacks;
+
+  private final int pageSize;
+
+  private final MetadataStore metadataStore;
+
+  private final ListeningExecutorService executor;
+
+  private List<DeleteObjectsRequest.KeyVersion> keys;
+
+  private List<Path> paths;
+
+  private CompletableFuture<Void> deleteFuture;
+
+  private long filesDeleted;
+  private long extraFilesDeleted;
+
+  /**
+   * Constructor.
+   * @param context store context
+   * @param status  pre-fetched source status
+   * @param recursive recursive delete?
+   * @param callbacks callback provider
+   * @param pageSize number of entries in a page
+   */
+  public DeleteOperation(final StoreContext context,
+      final S3AFileStatus status,
+      final boolean recursive,
+      final OperationCallbacks callbacks, int pageSize) {
+
+    super(context);
+    this.status = status;
+    this.recursive = recursive;
+    this.callbacks = callbacks;
+    checkArgument(pageSize > 0
+        && pageSize <=InternalConstants.MAX_ENTRIES_TO_DELETE,
+        "page size out of range: %d", pageSize);
+    this.pageSize = pageSize;
+    metadataStore = context.getMetadataStore();
+    executor = context.createThrottledExecutor(2);
+  }
+
+  public long getFilesDeleted() {
+    return filesDeleted;
+  }
+
+  public long getExtraFilesDeleted() {
+    return extraFilesDeleted;
+  }
+
+  /**
+   * Delete a file or directory tree.
+   * This call does not create any fake parent directory; that is
+   * left to the caller.
+   * The actual delete call is done in a separate thread.
+   * Only one delete at a time is submitted, however, to reduce the
+   * complexity of recovering from failures.
+   * <p>
+   * The DynamoDB store deletes paths in parallel itself, so that
+   * potentially slow part of the process is somewhat speeded up.
+   * The extra parallelization here is to list files from the store/DDB while
+   * that delete operation is in progress.
+   * <p>
+   * Note that DDB is not used for listing objects here, even if the
```

Comment:
yeah, will clarify. We still don't scan under empty dir markers. I was actually wondering today whether we should stop treating them as meaning "has no children", and stop shortcutting. This would let us avoid having to maybe create fake ones, delete fake ones, etc. would not be backwards compatible for sharing the same store across different client versions though